src/compiler/checker.ts

```diff
@@ -5696,6 +5696,29 @@ module ts {
             if (!links.resolvedType) {
                 links.resolvedType = checkExpression(node.expression);
 
+                // Disallow using static property in computedPropertyName because classDeclaration is binded lexically in ES6
+                // and its static property assignment will be emitted after classDeclaration.
+                // Therefore, using static property inside computedPropertyName will cause use-before-definition
+                // Example:
+                //  * TypeScript
+                //      class C {
+                //          static p = 10;
+                //          [C.p]() {}
+                //      }
+                //  * JavaScript
+                //      class C {
+                //          [C.p]() {}  // Use before definition error
+                //      }
+                //      C.p = 10;
+                if (languageVersion >= ScriptTarget.ES6 && links.resolvedSymbol) {

```

Comment:
Why not always have this rule, no matter if it's ES6 or ES5?  It seems very weird to allow it downlevel, and disallow it later.